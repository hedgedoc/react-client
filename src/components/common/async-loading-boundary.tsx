/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import type { PropsWithChildren } from 'react'
import React, { Fragment } from 'react'
import { Trans, useTranslation } from 'react-i18next'
import { WaitSpinner } from './wait-spinner/wait-spinner'
import { Alert } from 'react-bootstrap'

export interface AsyncLoadingBoundaryProps {
  loading: boolean
  error?: boolean
  componentName: string
}

/**
 * Indicates that a component currently loading or an error occurred.
 * It's meant to be used in combination with useAsync from react-use.
 *
 * @param loading Indicates that the component is currently loading. Setting this will show a spinner instead of the children.
 * @param error Indicates that an error occurred during the loading process. Setting this to any non-null value will show an error message instead of the children.
 * @param componentName The name of the component that is currently loading. It will be shown in the error message.
 * @param children The child {@link ReactElement elements} that are only shown if the component isn't in loading or error state
 */
export const AsyncLoadingBoundary: React.FC<PropsWithChildren<AsyncLoadingBoundaryProps>> = ({
  loading,
  error,
  componentName,
  children
}) => {
  useTranslation()
  if (error === true) {
    return (
      <Alert variant={'danger'}>
        <Trans i18nKey={'common.errorWhileLoading'} values={{ name: componentName }} />
      </Alert>
    )
  } else if (loading) {
    return <WaitSpinner />
  } else {
    return <Fragment>{children}</Fragment>
  }
}
