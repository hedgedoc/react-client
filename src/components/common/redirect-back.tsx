/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import React, { useEffect } from 'react'
import { useRouter } from 'next/router'

/**
 * Redirects the user back to the previous URL.
 */
export const RedirectBack: React.FC = () => {
  const router = useRouter()

  useEffect(() => {
    router.back()
  })

  return <span>Redirecting back</span>
}
