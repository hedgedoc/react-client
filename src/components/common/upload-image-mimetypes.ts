/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

export const supportedMimeTypes: string[] = [
  'image/apng',
  'image/bmp',
  'image/gif',
  'image/heif',
  'image/heic',
  'image/heif-sequence',
  'image/heic-sequence',
  'image/jpeg',
  'image/png',
  'image/svg+xml',
  'image/tiff',
  'image/webp'
]

export const acceptedMimeTypes = supportedMimeTypes.join(', ')
