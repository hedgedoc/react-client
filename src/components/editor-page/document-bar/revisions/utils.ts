/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { Revision } from '../../../../api/revisions/types'
import { getUserById } from '../../../../api/users'
import { UserResponse } from '../../../../api/users/types'
import { download } from '../../../common/download/download'
import { Logger } from '../../../../utils/logger'

const log = new Logger('RevisionsUtils')

export const downloadRevision = (noteId: string, revision: Revision | null): void => {
  if (!revision) {
    return
  }
  download(revision.content, `${noteId}-${revision.timestamp}.md`, 'text/markdown')
}

export const getUserDataForRevision = (authors: string[]): UserResponse[] => {
  const users: UserResponse[] = []
  authors.forEach((author, index) => {
    if (index > 9) {
      return
    }
    getUserById(author)
      .then((userData) => {
        users.push(userData)
      })
      .catch((error: Error) => log.error(error))
  })
  return users
}
