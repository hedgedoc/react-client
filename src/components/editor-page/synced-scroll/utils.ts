/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import type { LineMarkerPosition } from '../../markdown-renderer/markdown-extension/linemarker/types'

/**
 * Finds the surrounding {@link LineMarkerPosition line markers} from a list of line markers and the current line.
 *
 * @param lineMarks The list of line markers
 * @param lineNumber The current line
 * @return An object containing the last line mark before the current line and the next line mark after the current line
 * if applicable and undefined for each field otherwise
 */
export const findLineMarks = (
  lineMarks: LineMarkerPosition[],
  lineNumber: number
): { lastMarkBefore: LineMarkerPosition | undefined; firstMarkAfter: LineMarkerPosition | undefined } => {
  let lastMarkBefore
  let firstMarkAfter
  for (let i = 0; i < lineMarks.length; i++) {
    const currentMark = lineMarks[i]
    if (!currentMark) {
      continue
    }

    if (currentMark.line <= lineNumber) {
      lastMarkBefore = currentMark
    }
    if (currentMark.line > lineNumber) {
      firstMarkAfter = currentMark
    }
    if (!!firstMarkAfter && !!lastMarkBefore) {
      break
    }
  }
  return {
    lastMarkBefore,
    firstMarkAfter
  }
}
