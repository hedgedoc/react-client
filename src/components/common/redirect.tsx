/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import Link from 'next/link'
import React, { useEffect } from 'react'
import { useRouter } from 'next/router'
import { Logger } from '../../utils/logger'

export interface RedirectProps {
  to: string
}

const logger = new Logger('Redirect')

/**
 * Redirects the user to another URL. Can be external or internal.
 *
 * @param to The target URL
 */
export const Redirect: React.FC<RedirectProps> = ({ to }) => {
  const router = useRouter()

  useEffect(() => {
    router.push(to).catch((error: Error) => {
      logger.error(`Error while redirecting to ${to}`, error)
    })
  })

  return (
    <span>
      Redirecting to{' '}
      <Link href={to}>
        <a>{to}</a>
      </Link>
    </span>
  )
}
