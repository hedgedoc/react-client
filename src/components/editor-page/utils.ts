/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { isClientSideRendering } from '../../utils/is-client-side-rendering'

export const isMac: () => boolean = () => isClientSideRendering() && navigator.platform.toLowerCase().includes('mac')
