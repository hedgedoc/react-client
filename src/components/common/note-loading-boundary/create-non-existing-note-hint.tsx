/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { Trans, useTranslation } from 'react-i18next'
import React, { useCallback } from 'react'
import { Alert, Button } from 'react-bootstrap'
import { useSingleStringUrlParameter } from '../../../hooks/common/use-single-string-url-parameter'
import { createNoteWithPrimaryAlias } from '../../../api/notes'
import { useAsyncFn } from 'react-use'
import { ShowIf } from '../show-if/show-if'
import { Redirect } from '../redirect'
import { ForkAwesomeIcon } from '../fork-awesome/fork-awesome-icon'
import { testId } from '../../../utils/test-id'

/**
 * Shows a button that creates an empty note with the alias from the current window URL.
 * When the button was clicked it also shows the progress.
 */
export const CreateNonExistingNoteHint: React.FC = () => {
  useTranslation()
  const noteIdFromUrl = useSingleStringUrlParameter('noteId', undefined)

  const [returnState, createNote] = useAsyncFn(async () => {
    if (noteIdFromUrl === undefined) {
      throw new Error('Note id not set')
    }
    return await createNoteWithPrimaryAlias('', noteIdFromUrl)
  }, [noteIdFromUrl])

  const onClickHandler = useCallback(() => {
    void createNote()
  }, [createNote])

  if (noteIdFromUrl === undefined) {
    return null
  } else if (returnState.value) {
    return <Redirect to={`/n/${returnState.value.metadata.primaryAddress}`} />
  } else if (returnState.loading) {
    return (
      <div className={'d-flex flex-column mt-5'} {...testId('loadingMessage')}>
        <b>
          <ForkAwesomeIcon icon={'spinner'} className={'fa-spin mr-2'} />
          <Trans i18nKey={'noteLoadingBoundary.createNote.creating'} />
        </b>
      </div>
    )
  } else if (returnState.error !== undefined) {
    return (
      <Alert variant={'danger'} {...testId('failedMessage')} className={'mt-5'}>
        <ForkAwesomeIcon icon={'exclamation-triangle'} className={'mr-1'} />
        <Trans i18nKey={'noteLoadingBoundary.createNote.error'} />
      </Alert>
    )
  } else {
    return (
      <div className={'d-flex flex-column mt-5'}>
        <b>
          <Trans i18nKey={'noteLoadingBoundary.createNote.question'} values={{ aliasName: noteIdFromUrl }} />
        </b>
        <div className={'mt-3'}>
          <Button
            autoFocus
            type='submit'
            variant='primary'
            className='mx-2'
            onClick={onClickHandler}
            {...testId('createNoteButton')}>
            <ShowIf condition={returnState.loading}>
              <ForkAwesomeIcon icon={'spinner'} className={'fa-spin mr-2'} />
            </ShowIf>
            <Trans i18nKey={'noteLoadingBoundary.createNote.create'} />
          </Button>
        </div>
      </div>
    )
  }
}
