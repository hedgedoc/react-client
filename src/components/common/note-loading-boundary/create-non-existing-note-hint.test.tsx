/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import * as useSingleStringUrlParameterModule from '../../../hooks/common/use-single-string-url-parameter'
import { mockI18n } from '../../markdown-renderer/test-utils/mock-i18n'
import { act, render, screen } from '@testing-library/react'
import { CreateNonExistingNoteHint } from './create-non-existing-note-hint'
import * as createNoteWithPrimaryAliasModule from '../../../api/notes'
import type { Note, NoteMetadata } from '../../../api/notes/types'
import { Mock } from 'ts-mockery'
import { waitForOtherPromisesToFinish } from '../../../utils/wait-for-other-promises-to-finish'

describe('create non existing note hint', () => {
  const mockedNoteId = 'mockedNoteId'

  const mockGetNoteIdQueryParameter = () => {
    const expectedQueryParameter = 'noteId'
    jest.spyOn(useSingleStringUrlParameterModule, 'useSingleStringUrlParameter').mockImplementation((parameter) => {
      expect(parameter).toBe(expectedQueryParameter)
      return mockedNoteId
    })
  }

  const mockCreateNoteWithPrimaryAlias = () => {
    jest
      .spyOn(createNoteWithPrimaryAliasModule, 'createNoteWithPrimaryAlias')
      .mockImplementation(async (markdown, primaryAlias): Promise<Note> => {
        expect(markdown).toBe('')
        expect(primaryAlias).toBe(mockedNoteId)
        const metadata: NoteMetadata = Mock.of<NoteMetadata>({ primaryAddress: 'mockedPrimaryAlias' })
        await waitForOtherPromisesToFinish()
        return Mock.of<Note>({ metadata })
      })
  }

  const mockFailingCreateNoteWithPrimaryAlias = () => {
    jest
      .spyOn(createNoteWithPrimaryAliasModule, 'createNoteWithPrimaryAlias')
      .mockImplementation(async (markdown, primaryAlias): Promise<Note> => {
        expect(markdown).toBe('')
        expect(primaryAlias).toBe(mockedNoteId)
        await waitForOtherPromisesToFinish()
        throw new Error("couldn't create note")
      })
  }

  beforeAll(async () => {
    await mockI18n()
  })

  afterEach(() => {
    jest.resetAllMocks()
    jest.resetModules()
  })

  beforeEach(() => {
    mockGetNoteIdQueryParameter()
  })

  it('renders an button as initial state', async () => {
    mockCreateNoteWithPrimaryAlias()
    const view = render(<CreateNonExistingNoteHint></CreateNonExistingNoteHint>)
    await screen.findByTestId('createNoteMessage')
    expect(view.container).toMatchSnapshot()
  })

  it('renders a waiting message when button is clicked', async () => {
    mockCreateNoteWithPrimaryAlias()
    const view = render(<CreateNonExistingNoteHint></CreateNonExistingNoteHint>)
    const button = await screen.findByTestId('createNoteButton')
    act(() => {
      button.click()
    })
    await screen.findByTestId('loadingMessage')
    expect(view.container).toMatchSnapshot()
  })

  it('redirects when the note has been created', async () => {
    mockCreateNoteWithPrimaryAlias()
    const view = render(<CreateNonExistingNoteHint></CreateNonExistingNoteHint>)
    const button = await screen.findByTestId('createNoteButton')
    act(() => {
      button.click()
    })
    await screen.findByTestId('redirect')
    expect(view.container).toMatchSnapshot()
  })

  it("shows an error message if note couldn't be created", async () => {
    mockFailingCreateNoteWithPrimaryAlias()
    const view = render(<CreateNonExistingNoteHint></CreateNonExistingNoteHint>)
    const button = await screen.findByTestId('createNoteButton')
    act(() => {
      button.click()
    })
    await screen.findByTestId('failedMessage')
    expect(view.container).toMatchSnapshot()
  })
})
