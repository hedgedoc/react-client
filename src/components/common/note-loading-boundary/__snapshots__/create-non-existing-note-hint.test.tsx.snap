// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create non existing note hint redirects when the note has been created 1`] = `
<div>
  <span
    data-testid="redirect"
  >
    Redirecting to
     
    <a
      href="/n/mockedPrimaryAlias"
    >
      /n/mockedPrimaryAlias
    </a>
  </span>
</div>
`;

exports[`create non existing note hint renders a waiting message when button is clicked 1`] = `
<div>
  <div
    class="fade mt-5 alert alert-info show"
    data-testid="loadingMessage"
    role="alert"
  >
    <i
      class="fa  fa-spinner fa-spin mr-2 "
    />
    noteLoadingBoundary.createNote.creating
  </div>
</div>
`;

exports[`create non existing note hint renders an button as initial state 1`] = `
<div>
  <div
    class="fade mt-5 alert alert-info show"
    data-testid="createNoteMessage"
    role="alert"
  >
    <span>
      noteLoadingBoundary.createNote.question
    </span>
    <div
      class="mt-3"
    >
      <button
        class="mx-2 btn btn-primary"
        data-testid="createNoteButton"
        type="submit"
      >
        noteLoadingBoundary.createNote.create
      </button>
    </div>
  </div>
</div>
`;

exports[`create non existing note hint shows an error message if note couldn't be created 1`] = `
<div>
  <div
    class="fade mt-5 alert alert-danger show"
    data-testid="failedMessage"
    role="alert"
  >
    <i
      class="fa  fa-exclamation-triangle mr-1 "
    />
    noteLoadingBoundary.createNote.error
  </div>
</div>
`;
